### ISSUE TITLE
Improve front end performance of admin bar

### ISSUE THREAD

**Issue by dboulet** [backdrop/backdrop-issues#902]
Hi Nate, Jen and everyone,

One of the first things that I notice when using Backdrop is the janky way that the admin bar appears on the screen as the page loads. The way that things jump around as the bar is loaded is really jarring—not so much on fast page loads, but very perceivably on a slow connection.

I recreated the loading of the admin bar in slow mo using an animated GIF, to illustrated what I’m talking about:

![backdrop-admin-bar-load](https://cloud.githubusercontent.com/assets/208615/7442509/066e4f44-f0da-11e4-8f12-ee36d820bc58.gif)

There are a few issues that I can see, related to the admin bar’s over-dependance on JavaScript:
1. A top margin is applied to the page with JavaScript, to make room for the bar. Since this happens only after the page has loaded, the user sees the entire page jump down as the margin is applied.
2. The menus in the admin bar are made responsive using JavaScript. The most basic version of the menus are loaded first, which is then expanded into multiple drop downs only if there is enough space. Again, this is done after the page has loaded and the entire transformation is perceivable to the user.
3. The entire admin bar is cached on the client side and loaded onto the page using JavaScript. I’m assuming that this is done to save on bandwidth and server-side processing. It’s clever, but a bad idea for two reasons:
   - Bad perceived performance: the bar is only added to the page after the page has loaded, which means it’s completely missing as the page loads.
   - A user without JavaScript gets nothing. Missing JS could be caused by any number of reasons (JS turned off/blocked, slow connection/time out).

I would suggest a few solutions:
1. Add the top body margin on the server-side. Easily done by adding a class in `admin_bar_preprocess_page()`.
2. Make the bar responsive using media queries in CSS. This could be a little challenging since the width of the menus can change depending on chosen options and language, but not impossible.
3. Get rid of the client-side caching, in favour of server-side caching. Would sacrifice reduced page weight and bandwidth usage, but improves client-side page load and provides non-JS fallback.

This would allow us to remove much of the JavaScript from Administration Bar module. I’d be interested to know what you think.

**Comment by quicksketch** [comment]
Hi @dboulet, lots of good suggestions here. I think you're right that we can find several ways to improve the loading of the admin bar, though I think some of the suggestions are going after more difficult or trade-off implementations.

In speeding up the admin bar perceived performance, I'd like try to approach it with these specifications:

- The loading of the admin bar should not significantly impact the loading of the rest of the page (server-side generation or client-side rendering).
- We need to maintain the responsive and touch-friendly behaviors.
- Non-JS fallback is not a requirement.

I don't think we need to be painting JS as the bad-guy in the implementation. Although the loading of the menu is visible to the user, I definitely like that the menu is cached and displayed client-side (and after page load). I think it's preferable that we load the actual page content first, then add the admin bar afterwards, rather than holding up the entire page load on the admin bar. A lot of the JS is also more necessary than it may appear, for example the responsiveness is more complex than a media query would allow. And we need to use JS to have a proper touch-interface because touch devices have different implementations of `:hover` in CSS, so we need to bind both `onhover` and `ontouch` manually. And although the admin bar requires JS, in this situation, I think it's perfectly acceptable to make this a JS-only feature. It's an additive operation, and one of the advantages of loading it client-side makes is so a no-JS behavior is simply absent, rather than a half-working (at best) bar.

Here are some other options for speeding up the rendering, that might solve most of the visible problem without moving to a server-side only implementation:

- We can try to load the admin bar earlier or simultaneously instead of at the end of the page request.
- Eliminate the AJAX request by using localstorage if available (not sure about this).
- We can do the initial calculations for responsiveness off-screen and put the bar in place when ready. Or use media queries to guess the most likely initial starting position (not sure if that would work either).

**Comment by quicksketch** [comment]
> Eliminate the AJAX request by using localstorage if available (not sure about this).

Ah, I remember why we're not doing this. The AJAX request is *already cached* by the browser. Because the admin bar HTTP response sets a long cache lifetime, it's only loaded once and then all subsequent requests it's already in place. I usually have "Disable caches" turned on when the developer tools are open in my browser, but turning back on caching shows that the AJAX request is skipped on all page loads except the first one.

Rendering is still noticeable, but already the AJAX-request has no impact after the first page load. I think the combination of loading the menu earlier in the page process and doing calculations before showing the menu or using media queries to set "defaults" would help with the rendering time.

**Comment by dboulet** [comment]
Thanks for the feedback, @quicksketch. If you’re set on keeping the client-side caching, that’s totally fine—there are definitely things that we can still do to improve the loading time of the bar, and I like the suggestions that you made.

Making an earlier request for the admin bar markup is definitely a good idea, there’s no reason to wait until the document has loaded.

I still think it’s weird to have the admin bar be completely dependant on JavaScript. Most of the bar could be seen as a non-crucial enhancement, but it does contain some very important links, like **Log out**. Maybe there could be a slightly more progressively-enhanced way to go about it? One idea would be to output a placeholder toolbar which could be replaced using JavaScript. This would at least give the browser something to render while it waits for JavaScript to complete, and would help with page jumpiness. The placeholder could be something as simple as a super-basic toolbar containing only a “Log out” link.

**Comment by quicksketch** [comment]
Super, thanks for the PR @dboulet! I've been traveling the past few days but I'll take a look at this tomorrow. Thanks!

**Comment by quicksketch** [comment]
I've merged in the first PR at https://github.com/backdrop/backdrop/pull/865. Great job @dboulet! From the looks of things, we should still keep this issue open to allow for further improvements, correct? Any idea what improvement you'd like to see next?

**Comment by dboulet** [comment]
Thanks, @quicksketch. We can definitely follow this up with a few more improvements. Ideally, we want the admin bar to be in place before the page is rendered, so that the user doesn’t see anything jump after the page has loaded. Maybe the next step should be to see if we can make the AJAX request happen earlier in the page loading process, not having to wait until the document ready event has fired.

**Comment by quicksketch** [comment]
Sounds like a plan. I can't remember any reason why this behavior would need to come last. It may be the window.setTimeout() of 0 seconds that causes this to come last, as Admin Bar should be one of the first behaviors fired (starting with the letter A).

**Comment by dboulet** [comment]
The problem is not necessarily the order in which the behaviors are executed, but the fact that all behaviors are attached to jQuery’s `.ready()` function, which means that nothing is run until the DOM is fully loaded. In this case, we could detach the AJAX request from a behavior, and instead send the request immediately, as the script is being executed, to give it a head start while we wait for the DOM to load.

I looked into it a little more, and in reality it often doesn’t make a huge difference, since the AJAX response is cached most of the time anyway (a fact that you pointed out earlier). It might still be a good idea to optimize the uncached responses.

**Comment by quicksketch** [comment]
Thanks @dboulet! I tested your PR and added this special condition to Backdrop.attachBehaviors() to emulate a bunch of slow behaviors:

```js
Backdrop.attachBehaviors = function (context, settings) {
  context = context || document;
  settings = settings || Backdrop.settings;
  // Execute all of them.
  $.each(Backdrop.behaviors, function () {
    if ($.isFunction(this.attach)) {
      var me = this;
      window.setTimeout(function() {
        me.attach(context, settings);
      }, 500);
    }
  });
};
```

This revealed some interesting behavior, essentially running each behavior one at a time in slow-mo.

The admin bar *margin* was applied immediately, so fast it wasn't visible. However the admin bar itself was pokey to load, because it was still being called as a Drupal behavior. Instead of just applying the margin immediately, is there any reason why we couldn't apply the entire admin menu behavior at load time?

I changed `Backdrop.behaviors.adminBar.attach` to `Backdrop.adminBar.init` and called it at the bottom of the file (inside only the jQuery closure), and then the entire menu (AJAX request and all) is loaded before all other behaviors every time. Since the AJAX request is actually cached, the menu appears quite quickly, but unfortunately still gets stuck based on the number of available TCP connections. This makes the delay more noticeable in Chrome (which only allows 5 connections per origin) than Firefox (which allows 15).

I hadn't considered these connection limits previously. Even with the request cached, Chrome still prevents the display of cached content as if it required a TCP connection, even though it doesn't use one at all to display the cached copy. Very strange.

In any case, it's still less delay than before, loading admin menu first indeed makes the page appear less jumpy.

**Comment by dboulet** [comment]
Hey, @quicksketch. Unfortunately, we can’t insert the admin bar immediately because when we run the script in the `<head>`, the body element is not absolutely guaranteed to exist by the time the AJAX request completes. It works fine most of the time, but in certain cases the admin bar may end up not being inserted at all, because the there is no body element to insert into. As an example: another very heavy script could block the rendering of the page and delay the construction of the DOM, meanwhile our AJAX request response is cached and completes instantly.

To test this, take your `Backdrop.adminBar.init` example and add the following line just before the admin bar is inserted into the body:
```javascript
console.log($('body').length);
```
If you add a few heavy scripts after `admin_bar.js`, once in a while the above code will actually output 0.

This doesn‘t mean that we can’t send the AJAX request as early as possible, but we absolutely have to wait until the DOM is ready before inserting the response into the body.

On another note, your test results concerning the effect of browser connection limits on AJAX requests is interesting. Maybe it makes the case for opting to use localstorage, if available.

**Comment by quicksketch** [comment]
Thanks @dboulet! Very thoughtful. You're right that about 1 out of 20 times, I wouldn't get an admin bar at all. I thought it was a caching issue, but your explanation makes a lot more sense. Localstorage could indeed help this situation. Or perhaps we try the other suggestion and do the AJAX request early but only append it when the body tag loads. I can't think of a way to do the latter as of yet. I agree that localstorage may be a good solution here; I really hadn't expected cached requests to be affected by the connection limits.

**Comment by dboulet** [comment]
Added a new commit to my PR today, which allows the AJAX request to be sent as early as possible:

https://github.com/backdrop/backdrop/pull/876

If we can get that in, then I think that the only thing left to do would be to look into using localstorage to cache the admin bar in the browser.

**Comment by quicksketch** [comment]
Thanks @dboulet! We don't have the time to fit this in before 1.1.0, but I'll give this a review next week and we can fit it into the next bugfix release.

**Comment by dboulet** [comment]
Of course, thanks @quicksketch. There’s no big hurry with this, as it’s not a critical issue.

**Comment by quicksketch** [comment]
Hi @dboulet! Sorry about the delay on this one. Drupalcon came up and the 1.1.0 release and it's been pretty crazy ever since. I'm going to give this a shot.

**Comment by quicksketch** [comment]
Wow, this is great!! The new changes make the loading of the admin bar completely imperceptible. The white flash behind the bar is now eliminated and the menu appears to load faster than before. I can't detect any loading time at all. I've merged this into 1.x and 1.1.x. Let's open another issue if we decide that implementing localstorage would be worth it. Personally, I feel this is so fast now that adding it may be superfluous and we'd be better to leave the limited local storage space for end-user purposes.

**Comment by klonos** [comment]
> Wow, this is great!! The new changes make the loading of the admin bar completely imperceptible.

Same here on both a local dev environment as well as a live remote. Big difference! Thanx!!

**Comment by dboulet** [comment]
Thanks, @quicksketch. I agree, I think we’re done here.
